### PR TITLE
move includes around to lessen overall load

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: contrib/epee/include/storages/portable_storage_val_converters.h [hunk 1]
```diff
@@ -146,7 +146,7 @@ POP_WARNINGS
         if(std::all_of(from.begin(), from.end(), ::isdigit))
           to = boost::lexical_cast<uint64_t>(from);
         // MyMonero ISO 8061 timestamp (2017-05-06T16:27:06Z)
-        else if (std::regex_match (from, std::regex("\\d{4}-[01]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-5]\\dZ")))
+        else if (boost::regex_match (from, boost::regex("\\d{4}-[01]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-5]\\dZ")))
```
Comment (vtnerd) [inline contrib/epee/include/storages/portable_storage_val_converters.h]: Why was this changed to `boost::regex`?

File: contrib/epee/include/net/net_utils_base.h [hunk 2]
```diff
@@ -45,6 +42,7 @@
 #define MAKE_IP( a1, a2, a3, a4 )	(a1|(a2<<8)|(a3<<16)|(a4<<24))
 #endif
 
+namespace boost { namespace asio { class io_service; } }
```
Comment (vtnerd) [inline contrib/epee/include/net/net_utils_base.h]: This needs to be reverted. Newer versions of boost have changed the name to `io_context` and then `typedef io_context io_service`. So there is no longer a `class io_service`. This _should_ happen in the next release (boost 1.66).

File: contrib/epee/include/string_tools.h [hunk 3]
```diff
@@ -353,7 +353,12 @@ POP_WARNINGS
 	//----------------------------------------------------------------------------
 	inline bool validate_hex(uint64_t length, const std::string& str)
 	{
-		return std::regex_match(str, std::regex("'^[0-9abcdefABCDEF]+$'")) && str.size() == length;
+		if (str.size() != length)
```
Comment (vtnerd) [inline contrib/epee/include/string_tools.h]: A cpp already exists for this file, might as well move this code to that at the same time.

File: contrib/epee/include/string_tools.h [hunk 4]
```diff
@@ -353,7 +353,12 @@ POP_WARNINGS
...
+			return false;
+		for (char c: str)
+			if (!strchr("0123456789abcdefABCDEF", c))
```
Comment (vtnerd) [inline contrib/epee/include/string_tools.h]: I think this is an improvement over regex (which is abused a bit in epee), but what about:
```c++
for (char c : str) {
    if (!((u8'0' <= c && c <= u8'9') || (u8'a' <= c && c <= u8'f') || (u8'A' <= c && c <= u8'F')) {
        return false;
    }
}
```
`strchr` is a bit more complex because its assumes that that input string can be quite large (and is not looking for ranges). The `u8` is also a pessimism to ensure that ASCII-compatible values are being used instead of implementation defined character set.

File: contrib/epee/src/net_utils_base.cpp [hunk 5]
```diff
@@ -56,5 +57,34 @@ namespace epee { namespace net_utils
...
+			address = ipv4_network_address{ip, port};
+			return true;
+		}
+		return false;
+	}
+
+  std::string print_connection_context(const connection_context_base& ctx)
+  {
+    std::stringstream ss;
+    ss << ctx.m_remote_address.str() << " " << epee::string_tools::get_str_from_guid_a(ctx.m_connection_id) << (ctx.m_is_income ? " INC":" OUT");
```
Comment (vtnerd) [inline contrib/epee/src/net_utils_base.cpp]: Neither of these functions appears to require a `std::stringstream`. I wouldn't necessarily change it now (additional testing, etc.), but something to mark off in the future ...

File: src/blockchain_utilities/blocksdat_file.h [hunk 6]
```diff
@@ -31,8 +31,8 @@
 #include <boost/iostreams/stream_buffer.hpp>
 #include <boost/iostreams/stream.hpp>
 #include <boost/iostreams/device/back_inserter.hpp>
-
 #include <boost/iostreams/filtering_streambuf.hpp>
+#include <boost/filesystem.hpp>
```
Comment (vtnerd) [inline src/blockchain_utilities/blocksdat_file.h]: `#include <boost/filesystem/path.hpp>` might work here too ...

File: src/blockchain_utilities/bootstrap_file.h [hunk 7]
```diff
@@ -31,8 +31,8 @@
 #include <boost/iostreams/stream_buffer.hpp>
 #include <boost/iostreams/stream.hpp>
 #include <boost/iostreams/device/back_inserter.hpp>
-
 #include <boost/iostreams/filtering_streambuf.hpp>
+#include <boost/filesystem.hpp>
```
Comment (vtnerd) [inline src/blockchain_utilities/bootstrap_file.h]: Another `#include <boost/filesystem/path.hpp>`.

File: contrib/epee/include/storages/portable_storage_val_converters.h (hunk 1 above)
Comment (moneromooo-monero) [reply contrib/epee/include/storages/portable_storage_val_converters.h]: All other code uses boost::regex, so it avoids including both.

File: contrib/epee/include/string_tools.h (hunk 4 above)
Comment (moneromooo-monero) [reply contrib/epee/include/string_tools.h]: I was thinking of using ctype.h (isxdigit) instead actually.

File: src/blockchain_utilities/blocksdat_file.h (hunk 6 above)
Comment (moneromooo-monero) [reply src/blockchain_utilities/blocksdat_file.h]: Also needs operations.hpp, but that's still a win.

**Comment by moneromooo-monero** [comment]
I switched to using isxdigit, which is canonical. Let me know if you think there's a possible problem (ie, locales).

**Comment by vtnerd** [comment]
`std::isxdigit` is even better. Locales do not affect it AFAIK.
